gstd/src/lock/mutex.rs

```diff
@@ -117,27 +116,59 @@ impl<T> Mutex<T> {
 /// [`lock`](Mutex::lock) method on [`Mutex`].
 pub struct MutexGuard<'a, T> {
     mutex: &'a Mutex<T>,
+    holder_msg_id: MessageId,
+}
+
+impl<'a, T> MutexGuard<'a, T> {
+    fn ensure_access_by_holder(&self) {
+        let current_msg_id = msg::id();
+        if self.holder_msg_id != current_msg_id {
+            panic!(
+                "Mutex guard held by message 0x{} is being accessed by message 0x{}",
+                hex::encode(self.holder_msg_id),
+                hex::encode(current_msg_id)
+            );
+        }
+    }
 }
 
 impl<'a, T> Drop for MutexGuard<'a, T> {
     fn drop(&mut self) {
+        self.ensure_access_by_holder();
         unsafe {
-            *self.mutex.locked.get() = None;
-            if let Some(message_id) = self.mutex.queue.dequeue() {
-                crate::exec::wake(message_id).expect("Failed to wake the message");
+            let locked_by = &mut *self.mutex.locked.get();
+            if let Some(owner_msg_id) = *locked_by {
```

Comment:
As for me `if let Some(a) = b { ... } else { panic!("...") }` can be replaced by `a = b.unwrap_or_else(|| panic!("..."))`